Can't create lambda when package.json lives in parent directory

### What happened?

I'm updating my pulumi repo to have a single package.json at the root and it's worked fine everywhere except when dealing with inline lambdas.



### Expected Behavior

`pulumi up` works fine

### Steps to reproduce

Folder structure:
```
package.json
package-lock.json
node_modules/
repro/
  index.ts
  Pulumi.yaml
  Pulumi.production.yaml
  tsconfig.json
```

package.json:
```json
{
    "name": "repro",
    "devDependencies": {
        "@types/aws-sdk": "^2.7.0",
        "@types/node": "^10.17.60"
    },
    "dependencies": {
        "@pulumi/aws": "^5.42.0"
    }
}
```

index.ts:
```ts
import * as aws from '@pulumi/aws'

const lambdaFunction = new aws.lambda.CallbackFunction('my-callback', {
  callback: async (_) => {
    return "foo"
  },
  runtime: 'nodejs16.x',
})
```

tsconfig.json:
```json
{
    "compilerOptions": {
        "strict": true,
        "outDir": "bin",
        "target": "es2016",
        "module": "commonjs",
        "moduleResolution": "node",
    },
    "files": [
        "index.ts"
    ]
}
```

Error upon running `pulumi up`:
```
Previewing update (production)

View in Browser (Ctrl+O): https://app.pulumi.com/Activepipe/repro/production/previews/6a8bf1c8-91dc-4fb1-88f5-9954f9f46e5f

     Type                 Name              Plan       Info
 +   pulumi:pulumi:Stack  repro-production  create     1 error


Diagnostics:
  pulumi:pulumi:Stack (repro-production):
    error: Running program '/Users/jesse/repos/pulumi-lambda-repro/documentdb-repro/' failed with an unhandled exception:
    Error: Cannot find module './wafregional'
    Require stack:
    - /Users/jesse/repos/pulumi-lambda-repro/documentdb-repro/node_modules/@pulumi/aws/index.js
    - /Users/jesse/repos/pulumi-lambda-repro/documentdb-repro/index.ts
    - /Users/jesse/repos/pulumi-lambda-repro/node_modules/@pulumi/pulumi/cmd/run/run.js
    - /Users/jesse/repos/pulumi-lambda-repro/node_modules/@pulumi/pulumi/cmd/run/index.js
        at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
        at Function.Module._load (node:internal/modules/cjs/loader:778:27)
        at Module.require (node:internal/modules/cjs/loader:1005:19)
        at require (node:internal/modules/cjs/helpers:102:18)
        at Object.<anonymous> (/Users/jesse/repos/pulumi-lambda-repro/documentdb-repro/node_modules/@pulumi/index.ts:288:1)
        at Module._compile (node:internal/modules/cjs/loader:1105:14)
        at Object.Module._extensions..js (node:internal/modules/cjs/loader:1159:10)
        at Module.load (node:internal/modules/cjs/loader:981:32)
        at Function.Module._load (node:internal/modules/cjs/loader:822:12)
        at Module.require (node:internal/modules/cjs/loader:1005:19) {
      code: 'MODULE_NOT_FOUND',
      requireStack: [
        '/Users/jesse/repos/pulumi-lambda-repro/documentdb-repro/node_modules/@pulumi/aws/index.js',
        '/Users/jesse/repos/pulumi-lambda-repro/documentdb-repro/index.ts',
        '/Users/jesse/repos/pulumi-lambda-repro/node_modules/@pulumi/pulumi/cmd/run/run.js',
        '/Users/jesse/repos/pulumi-lambda-repro/node_modules/@pulumi/pulumi/cmd/run/index.js'
      ]
    }
```


### Output of `pulumi about`

▶ pulumi about
CLI
Version      3.78.1
Go Version   go1.20.7
Go Compiler  gc

Plugins
NAME    VERSION
aws     5.42.0
nodejs  unknown

Host
OS       darwin
Version  12.5.1
Arch     x86_64

This project is written in nodejs: executable='/Users/jesse/.nvm/versions/node/v16.15.1/bin/node' version='v16.15.1'

Current Stack: Activepipe/repro/production

Found no resources associated with production

Found no pending operations associated with production

Backend
Name           pulumi.com
URL            https://app.pulumi.com/Activepipe
User           Activepipe
Organizations  Activepipe

Dependencies:
NAME            VERSION
@pulumi/aws     5.42.0
@types/aws-sdk  2.7.0
@types/node     10.17.60

Pulumi locates its logs in /var/folders/d_/3fryg6d934l7jxttrz1z21bh0000gn/T/ by default

### Additional context

Symlinking package.json, package-lock.json and node_modules doesn't help.

If I move package.json/package-lock.json into the same folder as `index.ts` and `npm i` it works fine, but my goal is to have a single package.json defined at the root of the repo for all pulumi projects to use.

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

--- a/changelog/pending/20230824--sdk-nodejs--when-using-closure-serializtion-lookup-package-json-up-from-current-working-directory-up-to-parent-directories-recursively.yaml
+++ b/changelog/pending/20230824--sdk-nodejs--when-using-closure-serializtion-lookup-package-json-up-from-current-working-directory-up-to-parent-directories-recursively.yaml
@@ -1,0 +1,4 @@
+changes:
+- type: fix
+  scope: sdk/nodejs
+  description: When using closure serializtion, lookup package.json up from current working directory up to parent directories recursively

--- a/sdk/nodejs/runtime/closure/codePaths.ts
+++ b/sdk/nodejs/runtime/closure/codePaths.ts
@@ -163,7 +163,7 @@ function isSubsumedByHigherPath(normalizedPath: string, normalizedPathSet: Set<s
 }
 
 /**
- * searchUp searches for and returns the parent directory path
+ * searchUp searches for and returns the first directory path
  * starting from a given directory that contains the given file to find.
  * Recursively searches up the directory tree until it finds the file or returns null
  * when it can't find anything.

--- a/sdk/nodejs/runtime/closure/codePaths.ts
+++ b/sdk/nodejs/runtime/closure/codePaths.ts
@@ -162,11 +162,13 @@ function isSubsumedByHigherPath(normalizedPath: string, normalizedPathSet: Set<s
     return false;
 }
 
-// parentDirectory searches for and returns the parent directory path
-// starting from a given directory that contains the given file to find.
-// recursively searches up the directory tree until it finds the file or returns null
-// when it can't find anything.
-function parentDirectory(currentDir: string, fileToFind: string): string | null {
+/**
+ * searchUp searches for and returns the parent directory path
+ * starting from a given directory that contains the given file to find.
+ * Recursively searches up the directory tree until it finds the file or returns null
+ * when it can't find anything.
+ * */
+function searchUp(currentDir: string, fileToFind: string): string | null {
     if (fs.existsSync(upath.join(currentDir, fileToFind))) {
         return currentDir;
     }
@@ -174,7 +176,7 @@ function parentDirectory(currentDir: string, fileToFind: string): string | null 
     if (currentDir === parentDir) {
         return null;
     }
-    return parentDirectory(parentDir, fileToFind);
+    return searchUp(parentDir, fileToFind);
 }
 
 // allFolders computes the set of package folders that are transitively required by the root
@@ -186,7 +188,7 @@ function allFoldersForPackages(
 ): Promise<Set<string>> {
     return new Promise((resolve, reject) => {
         // the working directory is the directory containing the package.json file
-        const workingDir = parentDirectory(".", "package.json");
+        const workingDir = searchUp(".", "package.json");
         if (workingDir === null) {
             // we couldn't find a directory containing package.json
             // searching up from the current directory

--- a/sdk/nodejs/runtime/closure/codePaths.ts
+++ b/sdk/nodejs/runtime/closure/codePaths.ts
@@ -162,6 +162,21 @@ function isSubsumedByHigherPath(normalizedPath: string, normalizedPathSet: Set<s
     return false;
 }
 
+// parentDirectory searches for and returns the parent directory path 
+// starting from a given directory that contains the given file to find.
+// recursively searches up the directory tree until it finds the file or returns null
+// when it can't find anything.
+function parentDirectory(currentDir: string, fileToFind: string): string | null {
+  if (fs.existsSync(upath.join(currentDir, fileToFind))) {
+      return currentDir;
+  }
+  const parentDir = upath.resolve(currentDir, '..');
+  if (currentDir === parentDir) {
+      return null;
+  }
+  return parentDirectory(parentDir, fileToFind);
+}
+
 // allFolders computes the set of package folders that are transitively required by the root
 // 'dependencies' node in the client's project.json file.
 function allFoldersForPackages(
@@ -170,7 +185,15 @@ function allFoldersForPackages(
     logResource: Resource | undefined,
 ): Promise<Set<string>> {
     return new Promise((resolve, reject) => {
-        readPackageTree(".", <any>undefined, (err: any, root: readPackageTree.Node) => {
+        // the working directory is the directory containing the package.json file
+        const workingDir = parentDirectory(".", "package.json");
+        if (workingDir === null) {
+            // we couldn't find a directory containing package.json
+            // searching up from the current directory
+            throw new ResourceError("Failed to find package.json.", logResource);
+        }
+        
+        readPackageTree(workingDir, <any>undefined, (err: any, root: readPackageTree.Node) => {
             try {
                 if (err) {
                     return reject(err);
@@ -191,7 +214,7 @@ function allFoldersForPackages(
 
                     // From: https://github.com/npm/read-package-tree/blob/5245c6e50d7f46ae65191782622ec75bbe80561d/rpt.js#L121
                     root.package = computeDependenciesDirectlyFromPackageFile(
-                        upath.join(root.realpath, "package.json"),
+                        upath.join(workingDir, "package.json"),
                         logResource,
                     );
                 }

--- a/sdk/nodejs/runtime/closure/codePaths.ts
+++ b/sdk/nodejs/runtime/closure/codePaths.ts
@@ -167,14 +167,14 @@ function isSubsumedByHigherPath(normalizedPath: string, normalizedPathSet: Set<s
 // recursively searches up the directory tree until it finds the file or returns null
 // when it can't find anything.
 function parentDirectory(currentDir: string, fileToFind: string): string | null {
-  if (fs.existsSync(upath.join(currentDir, fileToFind))) {
-      return currentDir;
-  }
-  const parentDir = upath.resolve(currentDir, "..");
-  if (currentDir === parentDir) {
-      return null;
-  }
-  return parentDirectory(parentDir, fileToFind);
+    if (fs.existsSync(upath.join(currentDir, fileToFind))) {
+        return currentDir;
+    }
+    const parentDir = upath.resolve(currentDir, "..");
+    if (currentDir === parentDir) {
+        return null;
+    }
+    return parentDirectory(parentDir, fileToFind);
 }
 
 // allFolders computes the set of package folders that are transitively required by the root

--- a/sdk/nodejs/runtime/closure/codePaths.ts
+++ b/sdk/nodejs/runtime/closure/codePaths.ts
@@ -162,7 +162,7 @@ function isSubsumedByHigherPath(normalizedPath: string, normalizedPathSet: Set<s
     return false;
 }
 
-// parentDirectory searches for and returns the parent directory path 
+// parentDirectory searches for and returns the parent directory path
 // starting from a given directory that contains the given file to find.
 // recursively searches up the directory tree until it finds the file or returns null
 // when it can't find anything.
@@ -170,7 +170,7 @@ function parentDirectory(currentDir: string, fileToFind: string): string | null 
   if (fs.existsSync(upath.join(currentDir, fileToFind))) {
       return currentDir;
   }
-  const parentDir = upath.resolve(currentDir, '..');
+  const parentDir = upath.resolve(currentDir, "..");
   if (currentDir === parentDir) {
       return null;
   }
@@ -192,7 +192,6 @@ function allFoldersForPackages(
             // searching up from the current directory
             throw new ResourceError("Failed to find package.json.", logResource);
         }
-        
         readPackageTree(workingDir, <any>undefined, (err: any, root: readPackageTree.Node) => {
             try {
                 if (err) {

